idx/memory/tag_query.go

```diff
@@ -171,247 +252,751 @@ func NewTagQuery(expressions []string, from int64) (TagQuery, error) {
+
+IDS:
+	for id := range idCh {
+		var def *idx.Archive
+		var ok bool
+
+		if def, ok = q.byId[id.String()]; !ok {
+			// should never happen because every ID in the tag index
+			// must be present in the byId lookup table
+			corruptIndex.Inc()
+			log.Error(3, "memory-idx: ID %q is in tag index but not in the byId lookup table", id.String())
+			continue
+		}
+
+		// generate a set of all tags of the current metric that satisfy the
+		// tag filter condition
+		metricTags := make(map[string]struct{}, 0)
+		for _, tag := range def.Tags {
+			equal := strings.Index(tag, "=")
+			if equal < 0 {
+				corruptIndex.Inc()
+				log.Error(3, "memory-idx: tag is in index, but does not contain '=' sign: %s", tag)
+				continue
+			}
+
+			key := tag[:equal]
+			// this tag has already been pushed into tagCh, so we can stop evaluating
+			if _, ok := resultsCache[key]; ok {
+				continue
+			}
+
+			if q.filterTagBy == PREFIX_TAG {
+				if !strings.HasPrefix(key, q.tagPrefix) {
+					continue
+				}
+			} else if q.filterTagBy == MATCH_TAG {
+				if _, ok := q.tagMatch.missCache.Load(key); ok || !q.tagMatch.value.MatchString(tag) {
+					if !ok {
+						q.tagMatch.missCache.Store(key, struct{}{})
+					}
+					continue
+				}
+			}
+			metricTags[key] = struct{}{}
+		}
+
+		if matchName {
+			if _, ok := resultsCache["name"]; !ok {
+				metricTags["name"] = struct{}{}
+			}
+		}
+
+		// if some tags satisfy the current tag filter condition then we run
+		// the metric through all tag expression tests in order to decide
+		// whether those tags should be part of the final result set
+		if len(metricTags) > 0 {
+			if q.testByAllExpressions(id, def, matchName) {
+				for key := range metricTags {
+					select {
+					case tagCh <- key:
+					case <-stopCh:
+						// if execution of query has stopped because the max tag
+						// count has been reached then tagCh <- might block
+						// because that channel will not be consumed anymore. in
+						// that case the stop channel will have been closed so
+						// we so we exit here
+						break IDS
+					}
+					resultsCache[key] = struct{}{}
+				}
+			} else {
+				// check if we need to stop
+				select {
+				case <-stopCh:
+					break IDS
+				default:
+				}
+			}
+		}
+	}
+
+	q.wg.Done()
+}
+
+// determines whether the given tag prefix/tag match will match the special
+// tag "name". if it does, then we can omit some filtering because we know
+// that every metric has a name
+func (q *TagQuery) tagFilterMatchesName() bool {
+	matchName := false
+
+	if q.filterTagBy == PREFIX_TAG || q.startWith == PREFIX_TAG {
+		if strings.HasPrefix("name", q.tagPrefix) {
+			matchName = true
+		}
+	} else if q.filterTagBy == MATCH_TAG || q.startWith == MATCH_TAG {
+		if q.tagMatch.value.MatchString("name") {
+			matchName = true
+		}
+	} else {
+		// some tag queries might have no prefix specified yet, in this case
+		// we do not need to filter by the name
+		matchName = true
```

Comment:
this is confusing and seems to contradict the function documentation.  I think this entire function needs to explain better what it's trying to optimize.

Reply:
i don't think it contradicts, but i added another comment to further explain with an example: https://github.com/grafana/metrictank/pull/779/commits/3f55f9c0598668ac7209a2d29c985805c529a5a2

Reply:
maybe i should rename that variable to something like `omitTagFilters`. because thats the effect it has in the end, the fact that names match is only the reason why we want to omit the tag filters.

Reply:
like that https://github.com/grafana/metrictank/pull/779/commits/e50ea1e16f5d087c26ac08b684cf106db383138d